engine/docker/docker.go

```diff
@@ -525,3 +539,32 @@ func (e *Docker) setProxyInDockerDaemon(ctx context.Context, pipelineConfig *spe
 		}
 	}
 }
+
+// softStop stops the container giving them a 30 seconds grace period. The signal sent by ContainerStop is SIGTERM.
+// After the grace period, the container is killed with SIGKILL.
+// After all the containers are stopped, they are removed only when the status is not "running" or "removing".
+func (e *Docker) softStop(ctx context.Context, name string) {
+	timeout := 30 * time.Second
+	if err := e.client.ContainerStop(ctx, name, &timeout); err != nil {
+		logrus.WithField("container", name).WithField("error", err).Warnln("failed to stop the container")
+	}
+
+	// Before removing the container we want to be sure that it's in a healthy state to be removed.
+	now := time.Now()
+	for {
+		if time.Since(now) > timeout {
+			break
+		}
+		time.Sleep(1 * time.Second)
+		containerStatus, err := e.client.ContainerInspect(ctx, name)
```

Comment:
@nassergonzalez one question - doesn't the `timeout` you pass in `ContainerStop` take care of this logic? And do we want to try the stop only once?

Reply:
No, so the timeout that we pass to container stop uses that to stop the container and if in 30 secs it was not kill, to kill it, but the call does not block so it returns immediately. I thought the same as you when i started to do my investigation and later on discovered that that was not the case.

And yes, i think that trying to stop once is enough. That will send the stop and the docker engine will kill it eventually. Checking the container inspect is to be sure that we do not try to remove the container while it's been removing or running state. If we do not have that, what happens is:
Docker sends the stop signal
Before the plugin can handle the stop signal
Docker removes the container

Reply:
got it